pkg/rehearse/jobs.go

```diff
@@ -188,18 +201,70 @@ func ExecuteJobs(toBeRehearsed map[string][]prowconfig.Presubmit, prNumber int,
 		}
 	}
 
-	if len(rehearsals) > 0 {
-		for _, job := range rehearsals {
-			created, err := submitRehearsal(job, refs, logger, pjclient)
-			if err != nil {
-				logger.WithError(err).Warn("Failed to execute a rehearsal presubmit")
-			} else {
-				logger.WithFields(pjutil.ProwJobFields(created)).Info("Submitted rehearsal prowjob")
-			}
-		}
-	} else {
+	if len(rehearsals) == 0 {
 		logger.Warn("No job rehearsals")
+		return nil
+	}
+	pjs := make(sets.String, len(rehearsals))
+	for _, job := range rehearsals {
+		created, err := submitRehearsal(job, refs, logger, pjclient)
+		if err != nil {
+			logger.WithError(err).Warn("Failed to execute a rehearsal presubmit")
+			continue
+		}
+		logger.WithFields(pjutil.ProwJobFields(created)).Info("Submitted rehearsal prowjob")
+		pjs.Insert(created.Name)
+	}
+	req, err := labels.NewRequirement(rehearseLabel, selection.Equals, []string{strconv.Itoa(prNumber)})
+	if err != nil {
+		return fmt.Errorf("failed to create label selector: %v", err)
+	}
+	selector := labels.NewSelector().Add(*req).String()
+	success, err := waitForJobs(pjs, selector, pjclient, logger)
```

Comment:
Here is what the documentation for the `ListOptions` struct that is passed to the `Watch` call says:

https://github.com/kubernetes/apimachinery/blob/master/pkg/apis/meta/v1/types.go#L344

```go
// When specified with a watch call, shows changes that occur after that particular version of a resource.
// Defaults to changes from the beginning of history.
// …
ResourceVersion string `json:"resourceVersion,omitempty" protobuf:"bytes,4,opt,name=resourceVersion"`
```

I have also observed this in practice: before the label filter was added, every time I ran the program it would retell the long Tale of the 5000 Prowjobs. Note that it includes finished jobs, and the `switch` statement in `waitForJobs` will handle those correctly.

All this — and the fact that it behaved as expected in all my tests — indicate that the code is correct, although I don't know for certain.

Reply:
Cool, thanks for explanation. I was reading into this a bit yesterday evening and from what I understood it really works like this.